### PR TITLE
DAH-1360: Upgrade Partners to React 18

### PR DESCRIPTION
**Pull request by chadbrokaw**
[DAH-1360](https://sfgovdt.jira.com/browse/DAH-1360)

[DAH-1360]: https://sfgovdt.jira.com/browse/DAH-1360?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

This PR touches virtually everything in the application. There is nothing specific to test except ensuring that everything looks okay and that the communication between the front and back end is working as expected. A few things this PR does:

- Upgrades the app from React 16 to 18
- Removes Enzyme based tests and replaces them with RTL
- Removes webpacker and replaces it with shakapacker
- Upgrades react-router-dom to version 6

### CODE REVIEW

**Comment by tallulahkay** [comment]
Were we going to include an environment variable for the secret_key_base values?

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,18 +1,16 @@
...
-    "jest.rootPath": "./",
-    "autoimport.sourceRoot": "./app/javascript",
-    "autoimport.semicolon": false,
-    "autoimport.absolute": true,
-}
\ No newline at end of file
+  "javascript.validate.enable": true,
+  "jest.disabledWorkspaceFolders": ["spec/javascript/e2e"],
+  "editor.codeActionsOnSave": {
+    // For ESLint
```
Comment (jimlin-sfgov) [inline .vscode/settings.json]: suggestion: comments are not officially supported in JSON, and this comment seems unnecessary.

File: Gemfile [hunk 2]
```diff
@@ -100,11 +100,11 @@ gem "slim-rails", "~> 3.1"
 # Added at 2017-10-25 00:30:33 -0700 by dkaplan:
 gem "hashie"
 
-# Added at 2017-10-29 23:29:10 -0700 by dkaplan:
```
Comment (jimlin-sfgov) [inline Gemfile:103]: thought: are these comments even necessary? We have git to record this information

File: app/javascript/components/applications/application_form/preferences/PreferenceForm.js [hunk 3]
```diff
@@ -64,10 +64,12 @@ const PreferenceForm = ({ i, name, form, listingPreferences, fullHousehold }) =>
   )
 
   // Set the hidden recordType DeveloperName value
-  if (!!selectedPreference && selectedPreference.lottery_preference) {
-    const prefName = selectedPreference.lottery_preference.name
-    form.change(`preferences[${i}].recordtype_developername`, recordTypeMap[prefName] || 'Custom')
-  }
+  React.useEffect(() => {
```
Comment (jimlin-sfgov) [inline app/javascript/components/applications/application_form/preferences/PreferenceForm.js:67]: thought: I've noticed that we use custom versions of `useEffect` and `useState` hooks, which are defined in `customHooks.js`. We might want to revisit this later to determine if we actually need the custom versions. Maybe at the time they were implemented, React hooks were not mature enough to meet our needs?

File: app/javascript/routes/GoogleAnalyticsTracker.js [hunk 4]
```diff
@@ -1,25 +1,18 @@
...
 const GoogleAnalyticsTracker = ({ mockWindow = null }) => {
-  // this can't be inlined into the render function because of the useState hook.
-  const LogNewPathsOnRender = ({ location }) => {
-    const [lastPath, setLastPath] = useState(null)
+  const location = useLocation()
+  const [lastPath, setLastPath] = useState(null)
+
+  const ga = !mockWindow ? window.ga : mockWindow.ga
 
+  React.useEffect(() => {
```
Comment (jimlin-sfgov) [inline app/javascript/routes/GoogleAnalyticsTracker.js:12]: chore: can you include a section on how to if google analytics is working, in the Review Instructions?

File: spec/javascript/components/lease_ups/LeaseUpApplicationsFilterContainer.test.js [hunk 5]
```diff
@@ -30,92 +25,77 @@ const getNode = (bulkCheckboxesState = {}) => (
...
-      })
-
-      test('should render the checkbox as checked=false', () => {
-        expect(wrapper.find(Checkbox).props().checked).toBeFalsy()
+        getScreen({ appId1: false, appId2: true })
       })
 
       test('should render the checkbox with indeterminate=true', () => {
-        expect(wrapper.find(Checkbox).props().indeterminate).toBeTruthy()
+        // Even though the checkbox is in the indeterminate state, it is still considered checked
```
Comment (jimlin-sfgov) [inline spec/javascript/components/lease_ups/LeaseUpApplicationsFilterContainer.test.js]: suggestion: make the assertion cleaner with [toBePartiallyChecked](https://github.com/testing-library/jest-dom/pull/249)

File: Gemfile (hunk 2 above)
Comment (chadbrokaw) [reply Gemfile:103]: Yeah, they seem like clutter to me... I'm good to remove them

File: app/javascript/components/applications/application_form/preferences/PreferenceForm.js (hunk 3 above)
Comment (chadbrokaw) [reply app/javascript/components/applications/application_form/preferences/PreferenceForm.js:67]: Good catch. There are likely some of those that are not needed. I created a ticket to investigate this further: https://sfgovdt.jira.com/browse/DAH-1943

File: spec/javascript/components/lease_ups/LeaseUpApplicationsFilterContainer.test.js (hunk 5 above)
Comment (chadbrokaw) [reply spec/javascript/components/lease_ups/LeaseUpApplicationsFilterContainer.test.js]: Cool! Love that

File: app/javascript/routes/GoogleAnalyticsTracker.js (hunk 4 above)
Comment (chadbrokaw) [reply app/javascript/routes/GoogleAnalyticsTracker.js:12]: Sure! I added a section for how to test as well as an image that proves that the system is correctly talking with GA

**Review by jimlin-sfgov — APPROVED** [review]
Looks good, great job finding and updating all the test code to make the update possible!

**Pull request merged**